### PR TITLE
[REEF-1210] Implement SuspendTask on the Driver

### PR DESCRIPTION
**Pull request by afchung**
JIRA:
  [REEF-1210] https://issues.apache.org/jira/browse/REEF-1210

Pull Request:
  Closes #

### CODE REVIEW

**Comment by markusweimer** [comment]
LGTM, but I've been burned on bridge code before. Anyone else wants to take a close look?

**Comment by jwang98052** [comment]
I will look at it.

File: lang/cs/Org.Apache.REEF.Driver/Bridge/Events/RunningTask.cs [hunk 1]
```diff
@@ -60,12 +60,12 @@ public void Send(byte[] message)
 
         public void Suspend(byte[] message)
         {
-            throw new NotImplementedException();
+            _runningTaskClr2Java.Suspend(message);
         }
 
         public void Suspend()
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Driver/Bridge/Events/RunningTask.cs]: Can those two be consolidated?

File: lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/RunningTaskBridge.java [hunk 2]
```diff
@@ -47,6 +47,14 @@ public void send(final byte[] message) {
     jrunningTask.send(message);
   }
 
+  public void suspend(final byte[] message) {
+    if (message != null) {
+      jrunningTask.suspend(message);
+    } else {
+      jrunningTask.suspend();
```
Comment (jwang98052) [inline lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/RunningTaskBridge.java]: If this is the only place that suspend() is called, may be you can only keep public void Suspend(byte[] message) in RunningTask

File: lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        private const string SuspendValidationMessage = "SuspendValidationMessage";
+        private const string CompletedValidationMessage = "CompletedValidationmessage";
+
+        public TestSuspendTask()
+        {
+            Init();
+        }
+
+        /// <summary>
+        /// Does a simple test of whether a context can be submitted on top of another context.
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs]: Can you provide more details of the test scenario in the summary?

File: lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+            {
+                _requestor = evaluatorRequestor;
+            }
+
+            public void OnNext(IDriverStarted value)
+            {
+                _requestor.Submit(_requestor.NewBuilder().Build());
+            }
+
+            public void OnNext(IActiveContext value)
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs]: add comment, first time the task is submitted on the context?

File: lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+            {
+                value.SubmitTask(GetTaskConfiguration());
+            }
+
+            public void OnNext(IAllocatedEvaluator value)
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs]: add comment, the first context?

File: lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+                value.ActiveContext.Dispose();
+            }
+
+            public void OnNext(ISuspendedTask value)
+            {
+                Logger.Log(Level.Warning, SuspendValidationMessage);
+                value.ActiveContext.SubmitTask(GetTaskConfiguration());
+            }
+
+            public void OnNext(IRunningTask value)
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs]: Add comment: the suspend the first task?

File: lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+                        .Build());
+            }
+
+            public void OnNext(ICompletedTask value)
+            {
+                Logger.Log(Level.Warning, CompletedValidationMessage);
+                value.ActiveContext.Dispose();
+            }
+
+            public void OnNext(ISuspendedTask value)
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs]: add comment - received after the first task is suspended?

File: lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+            {
+                Logger.Log(Level.Warning, SuspendValidationMessage);
+                value.ActiveContext.SubmitTask(GetTaskConfiguration());
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs]: Submit another task on the same context? add comment.

**Comment by jwang98052** [comment]
I completed a pass. It looks good with minor comments.

**Comment by afchung** [comment]
@markusweimer @jwang98052 Sorry for the late update, I'll be addressing the comments today.

File: lang/cs/Org.Apache.REEF.Driver/Bridge/Events/RunningTask.cs (hunk 1 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Bridge/Events/RunningTask.cs]: What's the reasoning behind the need for consolidation? One sends a message and one doesn't.

File: lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/RunningTaskBridge.java (hunk 2 above)
Comment (afchung) [inline lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/RunningTaskBridge.java]: `RunningTask` is a `public` API. The user may call it.

File: lang/cs/Org.Apache.REEF.Driver/Bridge/Events/RunningTask.cs (hunk 1 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Driver/Bridge/Events/RunningTask.cs]: Note that this is a `public` API. The user can make calls to both.

File: lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs (hunk 3 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSuspendTask.cs]: Sorry, this was the result of a copy/paste from using the skeleton of another test. Will fix.

**Comment by afchung** [comment]
@markusweimer @jwang98052 Updated, please have another look. Thanks!

File: lang/cs/Org.Apache.REEF.Driver/Bridge/Events/RunningTask.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Driver/Bridge/Events/RunningTask.cs]: User can always call Suspend(null) if the user doesn't want to send message. Suspend() is a special case of Suspend(byte[]). If one can represent another, we don't need so many varieties.

I am OK to keep both if you think that is more explicit.

**Comment by tcNickolas** [comment]
This code doesn't build for me

    Error	14	No overload for method 'ValidateMessageSuccessfullyLogged' takes 2 arguments	C:\reef\lang\cs\Org.Apache.REEF.Tests\Functional\Bridge\TestSuspendTask.cs	65	13	Org.Apache.REEF.Tests

Our new AppVeyor setup agrees with me :-) https://ci.appveyor.com/project/ApacheSoftwareFoundation/reef/build/0.15.0-SNAPSHOT.7-master

Please revert the check-in or fix it.

**Pull request closed without merging**